### PR TITLE
Fix #19438. SpriteRenderer SetRenderStateForSprite.

### LINKED ISSUE #19438: SpriteRenderer SetRenderStateForSprite  has a probability of getting the wrong sheetindex
## Issue Summary

The class SpriteRenderer‘s method    SetRenderStateForSprite  can get wrong secondarySheetIndex，will cause shader get wrong depth 
texture，will cause some bug like below image。

## Reproduction

add a litte code  to  function  SetRenderStateForSprite at last line  can vertiy this  bug like below

`	if (sheetIndex == secondarySheetIndex && ss != null && ss.SecondarySheet!=s.Sheet)
			{
				Console.WriteLine("Have Problem????");
			}
`

this bug will trigger when sheet and secondarysheet all cache miss, then secondarysheetindex will  error
## Expected behavior

when  secondarysheet!=null and sheet!=secondarySheet   ,   sheetindex need not equal to secondarySheetIndex

## Screenshots / Screen recordings / Replays

if  change the shader( combined.frag)  frag_color to  output  vTexSampler.t, ,will get below result.

so  vTexSampler.t have some error

bbb

### PR DESCRIPTION
**Pull request by anvilvapre**
See https://github.com/OpenRA/OpenRA/issues/19438.

Untested. 

Index for secondary sheet was and will be zero when when the sprite has no secondary data.

### CODE REVIEW

**Comment by anvilvapre** [comment]
This does not consider the situation where `sheet` and `secondarySheet` are equal. Can this happen?

File: OpenRA.Game/Graphics/SpriteRenderer.cs [hunk 1]
```diff
@@ -69,41 +69,57 @@ int2 SetRenderStateForSprite(Sprite s)
 
 			// Check if the sheet (or secondary data sheet) have already been mapped
 			var sheet = s.Sheet;
-			var sheetIndex = 0;
-			for (; sheetIndex < ns; sheetIndex++)
-				if (sheets[sheetIndex] == sheet)
+			var sheetIndex = -1;
+			var addCount = 2;
```
Comment (abcdefg30) [inline OpenRA.Game/Graphics/SpriteRenderer.cs:73]: Shouldn't this be 1 by default and only increased by one if we have secondary data?

**Comment by pchote** [comment]
> This does not consider the situation where `sheet` and `secondarySheet` are equal. Can this happen?

I believe so, yes.

File: OpenRA.Game/Graphics/SpriteRenderer.cs (hunk 1 above)
Comment (anvilvapre) [reply OpenRA.Game/Graphics/SpriteRenderer.cs:73]: I believe that was the bug. When both sheet and secondary sheet did not exist - only one new sheet was added both pointing to that same new sheet.

Comment (pchote) [reply OpenRA.Game/Graphics/SpriteRenderer.cs:73]: Was this an existing bug, or a mistake in this PR?

If it was an existing bug then this will probably fix https://github.com/OpenRA/ra2/issues/710.

Comment (anvilvapre) [reply OpenRA.Game/Graphics/SpriteRenderer.cs:73]: That was the existing bug that this fix tries to solve.

**Comment by pchote** [comment]
I'm finding it quite hard to follow the branches in the current PR. Would df10339f5c also solve this?

**Comment by anvilvapre** [comment]
> I'm finding it quite hard to follow the branches in the current PR. Would [df10339](https://github.com/OpenRA/OpenRA/commit/df10339f5c4eaab20df2fb75fa773a6bd35864b5) also solve this?

My initial version looked more like that. Just that you are then 'adding' in two places. When i now flush i just say both sheets were not found - add them.

**Comment by pchote** [comment]
You're right - my first attempt was not well thought out. I spent some more time with this and came up with #19563 as a more straight forward approach. If you are happy that one solves the problem, can we close this?
